File: python/array/concatenation_array.py
def concatenate_arrays(numbers: list[int]) -> list[int]:
    # Allocate a 2n array and fill both halves in a single pass:
    # position i gets numbers[i], position i+n also gets numbers[i].
    # Time: O(n)  Space: O(n)
    numbers_length: int = len(numbers)
    concatenated_array: list[int] = [0] * (2 * numbers_length)
    idx: int
    for idx in range(numbers_length):
        concatenated_array[idx] = numbers[idx]
        concatenated_array[idx + numbers_length] = numbers[idx]
    return concatenated_array


def main() -> None:
    numbers: list[int] = list(map(int, input().split(" ")))

    concatenated_array: list[int] = concatenate_arrays(numbers)
    print(" ".join(map(str, concatenated_array)))


if __name__ == "__main__":
    main()

File: rust/src/bin/array/concatenation_array.rs
/*
Test Cases
1 2 3 4
1 2
3 1 4 1 5 9 2 6

Output
1 2 3 4 1 2 3 4
1 2 1 2
3 1 4 1 5 9 2 6 3 1 4 1 5 9 2 6
*/

use std::io::{self, BufRead};

// Allocate a 2n Vec and copy numbers into both halves:
// positions [0, n) and positions [n, 2n) each receive the full slice.
// Time: O(n)  Space: O(n)
fn concatenate_arrays(numbers: &[i64]) -> Vec<i64> {
    let numbers_length: usize = numbers.len();
    let mut concatenated_array: Vec<i64> = vec![0; 2 * numbers_length];
    concatenated_array[..numbers_length].copy_from_slice(numbers);
    concatenated_array[numbers_length..].copy_from_slice(numbers);
    concatenated_array
}

fn main() {
    let stdin = io::stdin();
    let line = stdin.lock().lines().next().unwrap().unwrap();
    let numbers: Vec<i64> = line
        .split_whitespace()
        .map(|s| s.parse().unwrap())
        .collect();

    let result: Vec<i64> = concatenate_arrays(&numbers);
    let strs: Vec<String> = result.iter().map(|v| v.to_string()).collect();
    println!("{}", strs.join(" "));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_concatenate_arrays() {
        assert_eq!(
            concatenate_arrays(&[1, 2, 3, 4]),
            vec![1, 2, 3, 4, 1, 2, 3, 4]
        );
        assert_eq!(concatenate_arrays(&[1, 2]), vec![1, 2, 1, 2]);
        assert_eq!(
            concatenate_arrays(&[3, 1, 4, 1, 5, 9, 2, 6]),
            vec![3, 1, 4, 1, 5, 9, 2, 6, 3, 1, 4, 1, 5, 9, 2, 6]
        );
    }
}
